### PR TITLE
Refactor Azure bucket configs

### LINKED ISSUE #2412: Using Azure Blob Storage causes Pyroscope Pod to just hang
#### Describe the bug
We've configured Pyroscope to use Azure Blob Storage as an Object Store, by setting the following in it's config map:

account_key: <Storage Account Key>
account_name: <Storage Account Name>
container_name: <Container Name>

When the Pyroscope container then starts up, it just seems to hang, with the following output in the logs:

ts=2023-09-18T09:55:10.230162485Z caller=azure.go:138 level=debug msg="creating new Azure bucket connection" component=storage

The pod never shows as ready/running, due to the health-check failing.

#### To Reproduce

1. Configure Pyroscope's Config Map to use an Azure Storage Account with Name/Key and a Container Name
2. Start Pyroscope Pod in AKS

#### Expected behavior

Pyroscope should connect to the Blob Storage account and start it's processes, or if there's a failure, present it in it's logs.

#### Environment

- Infrastructure: Kubernetes (AKS)
- Deployment tool: Helm

### PR DESCRIPTION
**Pull request by simonswine**
This PR fixes #2412 #2680 

It does that by refactoring the config to align with mimir's azure config.

It also avoids retrying for too often and too long, by lowering MaxRetries and making sure retries are done without exponential backoff. 

This depends on #2741

### CODE REVIEW

File: pkg/objstore/providers/azure/bucket_client.go [hunk 1]
```diff
@@ -9,26 +9,30 @@ import (
...
 	bucketConfig.MaxRetries = cfg.MaxRetries
 	bucketConfig.UserAssignedID = cfg.UserAssignedID
 
-	// Thanos currently doesn't support passing the config as is, but expects a YAML,
-	// so we're going to serialize it.
-	serialized, err := yaml.Marshal(bucketConfig)
-	if err != nil {
-		return nil, err
+	// do not delay retries
+	bucketConfig.PipelineConfig.RetryDelay = -1
```
Comment (aleks-p) [inline pkg/objstore/providers/azure/bucket_client.go:30]: Perhaps a naive question, why do we want to change the default behavior here? From what I can see Mimir doesn't do this. If we are worried about retries taking too long, there is also MaxRetryDelay that could be set.

**Review by aleks-p — APPROVED** [review]
I left one question but it looks good to me overall!

File: pkg/objstore/providers/azure/bucket_client.go (hunk 1 above)
Comment (aleks-p) [reply pkg/objstore/providers/azure/bucket_client.go:30]: Without the changes from #2741 I get the hang on startup with those 2 example commands. This goes away when I apply #2741.

Testing this branch, I don't see a difference in behavior between the changes here and latest `main`. Commenting this line out (reenabling retry delay) doesn't seem to make a difference as well for me.

Comment (simonswine) [reply pkg/objstore/providers/azure/bucket_client.go:30]: This not something I see, I do get a hang on main. I also don't see any big effect on max retries:

```
// main f80352606c8e13a0bcc65805e66899e5e664c275 with #2741, hangs > 1 min

// the max retries change and RetryDelay=-1, fails  after 10s

// the mimir config change, fails after 2-3s

// this pr as of right now, fail after 2-3s
```

I still think all of those changes should be done. MaxRetries=20 is too high, so this expoential back of

**Pull request merged**